src/plugins/action_trim_responses/src/trimResponses.ts

```diff
@@ -1,63 +1,77 @@
 import databaseMethods from './databaseMethods'
-import isEqual from '@openmsupply/deep-comparison'
-
+const isEqual = require('deep-equal')
 interface Response {
   id: number
   code: string
+  template_element_id?: number
+  application_response_id?: number
+  review_response_link_id?: number
   value: { [key: string]: any }
   comment?: string
   decision?: string
-  time_created: any
+  time_updated: any
 }
 
-interface ResponsesByCode {
-  [key: string]: Response[]
+interface ResponsesById {
+  [key: number]: Response[]
 }
 
 module.exports['trimResponses'] = async function (input: any, DBConnect: any) {
   const db = databaseMethods(DBConnect)
 
-  console.log('Trimming unchanged duplicated responses...')
+  const { applicationId, reviewId, timestamp } = input
+  console.log(`Trimming unchanged duplicated ${reviewId ? 'Review' : 'Application'} responses...`)
 
   try {
-    const { applicationId, reviewId } = input
     // Get ALL responses associated with application OR review
     const responses = reviewId
       ? await db.getAllReviewResponses(reviewId)
       : await db.getAllApplicationResponses(applicationId)
 
-    // Create object of responses indexed by code, with response
-    // objects added to array for each code
-    const responsesByCode: ResponsesByCode = {}
-    responses.forEach((response: Response) => {
-      const { code } = response
-      if (reviewId) response.value = { comment: response.comment, decision: response.decision }
-      if (!(code in responsesByCode)) responsesByCode[code] = [response]
-      else responsesByCode[code].push(response)
-    })
+    const reviewLevel = responses?.[0]?.level
+
+    // Create object of indexed responses, with array of response
+    // objects for each indexed id
 
+    const groupByField = !reviewId
+      ? 'template_element_id'
+      : reviewLevel > 1
+      ? 'review_response_link_id'
+      : 'application_response_id'
+
+    const responsesById = groupResponses(responses, groupByField)
     const responsesToDelete: number[] = []
+    const responsesToUpdate: number[] = []
 
-    Object.values(responsesByCode).forEach((responseArray) => {
+    Object.values(responsesById).forEach((responseArray) => {
       if (responseArray.length === 1) return
```

Comment:
can you add an 'undefined' review_response to the test and see if it's deleted please ?

use case -> reviewer starts review, marks one question as declined, all others should be deleted

Reply:
Okay, that should be deleting NULL responses now: ef4d5dc0

Reply:
I think this will still not quite work if there is one review_response

Reply:
Yeah good point. Think it should be correct now: d0aad32f

Just to clarify -- if a response has no duplicate (i.e. it's perId array will have one element), that element's timestamp should be updated too.